### PR TITLE
Propagate attributes to and from product proposal

### PR DESCRIPTION
**Pull request by metas-rc**

### CODE REVIEW

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRow.java [hunk 1]
```diff
@@ -162,6 +172,8 @@ private ProductsProposalRow(
 		this.packingDescription = packingDescription;
 		this.packingMaterialId = packingMaterialId;
 		this.asiDescription = asiDescription != null ? asiDescription : ProductASIDescription.NONE;
+		this.asiId = asiId;
```
Comment (adi-stefan) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRow.java:175]: Do you still need this here?

Comment (metas-rc) [reply backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRow.java:175]: I do, I use it here: de.metas.ui.web.order.products_proposal.service.OrderLinesFromProductProposalsProducer#createAndInitHUPackingAware

Comment (metas-rc) [reply backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRow.java:175]: https://github.com/metasfresh/metasfresh/pull/16573/files#diff-2370797d2c7fc150160828601f4b71229a931ec0a29319910b66cf8d12c6f692R190

**Review by adi-stefan — APPROVED** [review]

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java [hunk 2]
```diff
@@ -453,9 +453,35 @@ public AttributeSetInstanceId addAttributes(@NonNull final AddAttributesRequest
...
+	@Override
+	public boolean asiValuesMatchOrEmpty(@NonNull final AttributeSetInstanceId asiId1, @NonNull final AttributeSetInstanceId asiId2)
+	{
+		final List<I_M_AttributeInstance> attributeInstances1 = attributeDAO.retrieveAttributeInstances(asiId1);
+
+		boolean matches = true;
+
+		for (final I_M_AttributeInstance attributeInstance1 : attributeInstances1)
+		{
+			final I_M_AttributeInstance attributeInstance2 = attributeDAO.retrieveAttributeInstance(asiId2, AttributeId.ofRepoId(attributeInstance1.getM_Attribute_ID()));
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java:471]: @metas-rc cc @adi-stefan  isn't this an SQL N+1 issue?
why we didn't retrieve all the AIs of second ASI all together?

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java [hunk 3]
```diff
@@ -453,9 +453,35 @@ public AttributeSetInstanceId addAttributes(@NonNull final AddAttributesRequest
...
 		request.getAttributeInstanceBasicInfos().forEach(attributeValue -> {
-			setAttributeInstanceValue(asiId, attributeValue.getAttributeCode(), attributeValue.getValue() );
+			setAttributeInstanceValue(asiId, attributeValue.getAttributeCode(), attributeValue.getValue());
 		});
 
 		return asiId;
 	}
+
+	@Override
+	public boolean asiValuesMatchOrEmpty(@NonNull final AttributeSetInstanceId asiId1, @NonNull final AttributeSetInstanceId asiId2)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java:463]: @metas-rc  @adi-stefan  why not prefixing the method with `is`?
also, i think the name is a bit ambiguous/misleading. Maybe something like `isCommonAttributesMatching`...

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java [hunk 4]
```diff
@@ -453,9 +453,35 @@ public AttributeSetInstanceId addAttributes(@NonNull final AddAttributesRequest
...
+		for (final I_M_AttributeInstance attributeInstance1 : attributeInstances1)
+		{
+			final I_M_AttributeInstance attributeInstance2 = attributeDAO.retrieveAttributeInstance(asiId2, AttributeId.ofRepoId(attributeInstance1.getM_Attribute_ID()));
+
+			if (attributeInstance2 == null)
+			{
+				continue;
+			}
+
+			if (!Objects.equals(attributeInstance1.getValue(), attributeInstance2.getValue()))
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java:478]: @metas-rc @adi-stefan  what about ValueNumber and ValueDate?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRowsLoader.java [hunk 5]
```diff
@@ -224,12 +226,16 @@ private ProductsProposalRow toProductsProposalRowOrNull(@NonNull final I_M_Produ
 				? packingMaterialsService.getDisplayName(packingMaterialId)
 				: TranslatableStrings.empty();
 
+		final Predicate<OrderLine> asiMatcher = orderLine -> attributeSetInstanceBL.asiValuesMatchOrEmpty(orderLine.getAsiId(), extractProductASI(record));
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRowsLoader.java:229]: @metas-rc @adi-stefan auch, now i see that we will load the record's ASI on each call.
IMHO asiValuesMatchOrEmpty method shall be changed to work with `ImmutableAttributeSet` which is loaded only once here.

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRowsLoader.java [hunk 6]
```diff
@@ -246,6 +252,11 @@ private ProductASIDescription extractProductASIDescription(final I_M_ProductPric
 		return ProductASIDescription.ofString(attributeSetInstanceBL.getASIDescriptionById(asiId));
 	}
 
+	private AttributeSetInstanceId extractProductASI(final I_M_ProductPrice record)
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/model/ProductsProposalRowsLoader.java:255]: @metas-rc  @adi-stefan  make it static?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderLine.java [hunk 7]
```diff
@@ -49,25 +50,31 @@ public class OrderLine
...
+
 	int qtyEnteredTU;
 
 	String description;
 
 	boolean isMatching(
 			@NonNull final ProductId productId,
-			@Nullable final HUPIItemProductId packingMaterialId)
+			@Nullable final HUPIItemProductId packingMaterialId,
+			@NonNull Predicate<OrderLine> asiMatcher)
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderLine.java:70]: @metas-rc @adi-stefan  IMHO makes no sense to have asiMatcher parameter here. You can simply call it from outside, no need to "call you to call me"

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderLinesFromProductProposalsProducer.java [hunk 8]
```diff
@@ -140,7 +139,7 @@ private void updateOrderLine(
 
 		huPackingAwareBL.prepareCopyFrom(rowPackingAware)
 				.overridePartner(false)
-				.asiCopyMode(ASICopyMode.CopyID) // because we just created the ASI
+				.asiCopyMode(ASICopyMode.Clone)
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderLinesFromProductProposalsProducer.java:142]: @metas-rc  @adi-stefan why changing from CopyID to Clone?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderLinesFromProductProposalsProducer.java [hunk 9]
```diff
@@ -188,7 +187,7 @@ private PlainHUPackingAware createAndInitHUPackingAware(
 		final UomId uomId = productBL.getStockUOMId(fromRow.getProductId());
 		huPackingAware.setProductId(fromRow.getProductId());
 		huPackingAware.setUomId(uomId);
-		// huPackingAware.setM_AttributeSetInstance_ID(...);
+		huPackingAware.setM_AttributeSetInstance_ID(AttributeSetInstanceId.toRepoId(fromRow.getAsiId()));
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderLinesFromProductProposalsProducer.java:190]: @metas-rc @adi-stefan  why `fromRow.getAsiId()` is not null?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderProductProposalsService.java [hunk 10]
```diff
@@ -110,6 +111,7 @@ private OrderLine toOrderLine(final I_C_OrderLine record)
 				.qtyEnteredCU(record.getQtyEntered())
 				.qtyEnteredTU(record.getQtyEnteredTU().intValue())
 				.description(record.getDescription())
+				.asiId(AttributeSetInstanceId.ofRepoId(record.getM_AttributeSetInstance_ID()))
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/products_proposal/service/OrderProductProposalsService.java:114]: @metas-rc @adi-stefan  IMHO we shall use ofRepoIdOrNone

**Review by teosarca — COMMENTED** [review]
@metas-rc @adi-stefan  pls check my comments

**Pull request merged**